nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+use with the pre-defined RPCs as shown below:
+
+- `Start(operation, input, callbackURL, cancelBeforeStarted)` - handler responds inline or returns a reference Id
+  ([`*`](#cancel-before-started))
+- `Cancel(reference_id)` - cancel operation by handler provided reference Id
+- `GetResult(reference_id)` - get operation result by handler provided reference Id
+- `Describe(reference_id)` - get operation status by handler provided reference Id
+- `DeliverResult(token)` - deliver a result of an async operation using a token in the Start request's callback URL
+
+The handler _should_ either sign or qualify caller-provided reference Ids when dealing with untrusted sources and
+multi-tenant applications.
+
+In order to support stateless clients, the handler may provide an additional set of APIs to resolve client input to a
+reference Id. This may be something that is built into the protocol or implemented ad-hoc. To support this at the
+protocol level, we would need to add the following APIs:
+
+- `Resolve(operation, input) -> reference_id`
+
+-- OR --
+
+- `Cancel(operation, input)` - cancel operation by handler derived Id
+- `GetResult(operation, input)` - get operation result by handler derived Id
+- `Describe(operation, input)` - get operation status by handler derived Id
+
+NOTE: This looks a lot like the APIs for the "client generated Ids" approach, except the ID is structured instead of a
+single string.
+
+#### Cancel before started
+
+A caller may cancel an operation before it obtains a reference Id, e.g. if it is uncertain that the operation has been
+already started due to crashes and timeouts. Because the caller doesn't have a referencable Id or knowledge of which
+input fields are used to derive the Id, it would need to pass in the entire Start call input.
+
+This is especially important when we want to automate cancellations of workflow initiated operations, Temporal server
+cannot derive the Id from opaque input.
+
+### Comparing the approaches
+
+#### Handler Id qualification
+
+With handler generated Ids, to support stateless clients, the handler would need to be able to map an operation and
+input to a reference Id. It would also need to qualify a reference Id or securely issue the reference Ids when callers
+cannot be trusted. With caller generated Ids, the handler would only need to qualify an Id based on the caller context
+(e.g by concatinating it with a tenant Id).
+
+#### Ids as inputs
+
+With the caller generated Id approach, if the Id is used for handler side addressability, the handler would need to
+parse it and merge with the request's input.
+
+#### Caller friendliness
+
+Shifting the responsibility for creating a handler addressable Id to the caller makes an API harder to use.
+
+If we can map arbitrary client Ids to handler addressable resources, the client generated Id approach is easiest to use
+from stateless clients. The SDK can generate deterministic Ids for calls started from workflows.
+
```

Comment:
This breaks down when you start to share the identifiers, so that others can GetResult(). The structure that makes sense for your program, could be very different to whoever created (and named) the identifier...

Reply:
I will add something to the proposal about sharing identifiers between different consumers (multi-tenant apps).